rfd/0160-kubernetes-operator-resource-versioning.md

```diff
@@ -0,0 +1,210 @@
+more details.
+
+### Pros
+
+- This is backward compatible as we can default to the existing versions when
+  the `teleportResourceVersion` field is not set. We can also continue to support the
+  existing CRD APIs.
+- The implementation cost of this solution is several orders of magnitude lower
+  than the alternatives. See [the Alternatives section](#alternatives).
+- This is the same versioning approach we are already using for the Terraform Provider.
+- This design doesn't add new failure modes, unlike solutions implying conversion webhooks.
+- Upgrading the resource version is easy for a Teleport user: they change both the
+  `teleportResourceVersion` and the affected `spec` fields simultaneously.
+
+### Cons
+
+- This design doesn't follow usual Kubernetes resource versioning.
+- Settling on `vX` can be confusing. See [the API evolution section](#api-evolution).
+- This design relies on the fact we don't radically change the resource structure.
+  See [the API evolution section](#api-evolution).
+
+### Security
+
+This design adds little changes to the current Teleport Kubernetes Operator
+security model. The only risk is users forgetting to set the `teleportResourceVersion`
+field, in this case we would default to the current version. In a future operator
+version, we can enforce the presence of this field if deemed necessary.
+
+### UX
+
+- This change is backward compatible and offers a straightforward migration
+  path when we introduce a new resource version.
+- The atypical versioning might be confusing for users as it doesn't follow
+  usual Kubernetes patterns.
+
+### API evolution
+
+This design unblocks us today, but does not cover cases in which we introduce a
+new resource version with a completely different structure with fields with
+different types. It is unclear if this can or will happen as this would very
+likely cause other issues in Teleport due to the way we handle resource versioning.
+
+If this were to happen, we would need to revise the solution from this RFD and
+introduce a new resource, or implement conversion hooks. We would also need to
+change the `resources.teleport.dev/vX` version to the current Teleport version.
+
+For example:
+- We implement this RFD in Teleport 15: the API is `resources.teleport.dev/v15`
+- We don't do any breaking changes before v20: the API is still `resources.teleport.dev/v15` for Teleport version 16, 17, 18 and 19.
+- In Teleport v20, we introduce a breaking change in Role. We introduce a new API `resoucres.teleport.dev/v20`.
+
+### Alternatives
+
+When writing this RFD, two alternatives were considered:
+
+#### Conversion hooks
+
+The Kubernetes-friendly approach would be to make the operator aware of how the
+resource is stored in Kubernetes, and do the conversions for every Kubernetes CR
+API call via webhooks.
+
+This alternative has been rejected because of the following reasons:
+- We currently handle resource conversion only from the old version to the newer versions,
```

Comment:
This solution can be delayed.  We can just return the role v5 with the payload of role v7 for now. It's something that's ok and it's less broken than our current version.

Adding a validation webhook would prevent users from creating invalid roles which is also a good thing because now users don't know if the role is correct until they inspect the events (not great as it forces operator to reconcile invalid roles)

Reply:
> it's less broken than our current version.

I'm afraid it will be more broken than the current version and the other alternatives. If we handle conversion at the operator level, we must also handle the default setting and the resource validation.

By moving those responsibilities to the operator, we will end up in a brain-split situation where both the operator and Teleport will set defaults. This can cause both to set different defaults, either because of a bug or a version mismatch.

I think it is better for the operator to stay as dumb as possible. Acting as Teleport and trying to set defaults for the sake of looking like other Kubernetes operators will very likely hurt us.

A saner conversion hook approach would be to handle every conversion on the Teleport side. However, I don't think this is viable from a latency, reliability and performance point of view.